qiskit_machine_learning/utils/adjust_num_qubits.py

```diff
@@ -0,0 +1,112 @@
+"""Helper functions to adjust number of qubits."""
+
+from typing import Tuple
+
+from qiskit import QuantumCircuit
+from qiskit.circuit.library import RealAmplitudes, ZFeatureMap, ZZFeatureMap
+
+from qiskit_machine_learning import QiskitMachineLearningError
+
+
+# pylint: disable=invalid-name
+def derive_num_qubits_feature_map_ansatz(
+    num_qubits: int = None, feature_map: QuantumCircuit = None, ansatz: QuantumCircuit = None
+) -> Tuple[int, QuantumCircuit, QuantumCircuit]:
+    """
+    Derives a correct number of qubits, feature map, and ansatz from the parameters.
+
+    If the number of qubits is not ``None``, then the feature map and ansatz are adjusted to this
+    number of qubits if required. If such an adjustment fails, an error is raised.
+    Also, if the feature map or ansatz or both are ``None``, then ``ZZFeatureMap`` and
+    ``RealAmplitudes`` are created respectively. If there's just one qubit, ``ZFeatureMap`` is
+    created instead.
+
+    If the number of qubits is ``None``, then the number of qubits is derived from the feature map
+    or ansatz. Both the feature map and ansatz in this case must have the same number of qubits.
+    If the number of qubits of the feature map is not the same as the number of qubits of
+    the ansatz, an error is raised. If only one of the feature map and ansatz are ``None``, then
+    ``ZZFeatureMap`` or ``RealAmplitudes`` are created respectively.
+
+    If all the parameters are none an error is raised.
+
+    Args:
+        num_qubits: Number of qubits.
+        feature_map: A feature map.
+        ansatz: An ansatz.
+
+    Returns:
+        A tuple of number of qubits, feature map, and ansatz. All are not none.
+
+    Raises:
+        QiskitMachineLearningError: If correct values can not be derived from the parameters.
+    """
+    # check num_qubits, feature_map, and ansatz
+    if num_qubits is None and feature_map is None and ansatz is None:
+        raise QiskitMachineLearningError(
+            "Need at least one of number of qubits, feature map, or ansatz!"
+        )
+
+    if num_qubits is not None:
+        if feature_map is not None:
+            if feature_map.num_qubits != num_qubits:
+                _adjust_num_qubits(feature_map, "feature map", num_qubits)
+        else:
+            if num_qubits == 1:
+                feature_map = ZFeatureMap(1)
+            else:
+                feature_map = ZZFeatureMap(num_qubits)
+        if ansatz is not None:
+            if ansatz.num_qubits != num_qubits:
+                _adjust_num_qubits(ansatz, "ansatz", num_qubits)
+        else:
+            ansatz = RealAmplitudes(num_qubits)
+    else:
+        if feature_map is not None and ansatz is not None:
+            if feature_map.num_qubits != ansatz.num_qubits:
+                raise QiskitMachineLearningError(
+                    f"Mismatching number of qubits in the feature map ({feature_map.num_qubits}) "
+                    f"and the ansatz ({ansatz.num_qubits})!"
```

Comment:
In this case, if I supply both, then I have to ensure they have matching number of qubits, by setting them both. I guess thats fine. In the docstring of this method there is ample description of all of this. But this function is not in the API ref, I don;t think we want to expose it right, its more an internal utility and if I look at VQC etc in this context (num_qubits, feature_map and anzazt args there) the explanation there and elsewhere in similar places is arguably a bit light. I am not sure as verbose as the docstring here is needed, but maybe it could have a little more around the behavior like this.

Reply:
- If you supply both, then I don't know which one has the correct number of qubits, so I can't set them both.
- I updated the docstrings in `TwoLayerQNN`, `VQC`, and `VQR` where the parameters are similar.
- Updated the type hints in these classes to be consistent. All classes should be reviewed to make use of the new style of type hints.